README.md
# BQN VS Code plugin

This VS Code extension provides language support for [BQN].

It is available on the [Open VSX Registry] and the [VS Code Marketplace].

## Setup
This extension has a focus on the [CBQN] implementation. The `REPLXX=1` option
must be enabled during building for best results.

### Configurable Settings
You can add these settings to your settings.json file to disable some of the default behaviours of this extension.
- `bqn.executablePath` - Path to the CBQN executable e.g.
  `/home/username/CBQN/BQN`.
- `bqn.saveBeforeLoadScript` - Whether vscode should save the script before
   loading it into the CBQN REPL.
- `bqn.sendToNewReplDelay` - How long vscode should wait before sending text to
   a repl.
- `bqn.executableSupportsReplxx` - Marks whether the executable set in
   `bqn.executablePath` is built with replxx support.
- `bqn.pendingBackslashBackgroundColor` - Style customization feature for
   backslash completion.
- `bqn.enableBackslashCompletion` - `true` by default. If you have an extension 
   that interferes with backslash completion (e.g. Vim keybindings), please set
   this to false and use a
   [system-wide method.](https://mlochbaum.github.io/BQN/editors/#system-wide)
- `bqn.enableHoverDocumentation` - Controls whether documentation will be
   displayed for a symbol on mouseover.

## Features

### Current

- Syntax highlighting.
- Backslash symbol completion (`\r` -> `↑`) using the [standard BQN keymap],
  like in the [online REPL].
- Symbol completion by function name (`take<tab>` -> `↑`).
- Help popups when hovering over glyphs.
- File, line, and selection execution in the integrated terminal (from @suhr and @mk12). This requires having a `bqn` executable in your PATH, or customizing the `bqn.executablePath` setting.

### Planned

- Block context highlighting for symbols like `𝕨` and `𝕣`.
- Inline code execution similar to [nvim-bqn].
- CBQN WASM interpreter bundled with the extension.
- Add a symbol table to the sidebar for clickable BQN symbols.

[BQN]: https://mlochbaum.github.io/BQN/
[CBQN]: https://github.com/dzaima/CBQN
[Open VSX Registry]: https://open-vsx.org/extension/razetime/bqn-language
[VS Code Marketplace]: https://marketplace.visualstudio.com/items?itemName=razetime.bqn-language
[Standard BQN keymap]: https://mlochbaum.github.io/BQN/keymap.html
[online REPL]: https://mlochbaum.github.io/BQN/try.html
[nvim-bqn]: https://git.sr.ht/~detegr/nvim-bqn


CHANGELOG.md
# Change Log

All notable changes to the "bqn-language" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]
- Block context highlighting for symbols like `𝕨` and `𝕣`
- inline code execution similar to [nvim-bqn](https://git.sr.ht/~detegr/nvim-bqn)
- CBQN WASM interpreter bundled with the extension
- BQNcrate search

## Released
- v0.1.6
  - Add BQN language file icon
  - Option to disable backslash completion
- v0.1.5
  - remove incomplete changes for bqncrate integration
- v0.1.3
  - Add Latin1 variable support
  - Correct TextMate classes for highlighting
  - Fix double-struck characters not being highlighted
  - Add in function name-based symbol completion
- v0.1.2
  - fix syntax highlighting problem with strands (`‿`)
- v0.1.1 (changes by @mk12)
  - Use bracketed paste mode
  - "Clear Imports" and "Profile Script" commands
  - Support multiple cursors
  - Highlight pending backslash
  - Documentation for symbols on hover
- v0.1.0 (changes by @mk12)
  * **Implement "Execute Selection"**.
  * **Rewrite the backslash command** to be more reliable.
  * **Create the REPL in the terminal panel**, not in an editor (which was annoying because it covered the current editor).
  * When executing code requires creating a new REPL, **delay for 500ms** (customizable) to wait for BQN to be ready, otherwise input gets messed up in my experience. Also delay 50ms before sending the newline after an `)ex` command for similar reasons.
  * **Start the REPL in the current file's directory**, rather than the project directory (default), so that imports work correctly.
  * For execution commands, **keep focus in the editor**, but for the "Create BQN REPL" command, focus the terminal.
  * When advancing to the next line, **skip over blank lines and comments**.
  * Use title case for "BQN: Convert To Symbol".
  * Remove unused codeLensProvider capability.
  * Remove semantic highlighting, which didn't work and was causing errors in the logs (seems to have been copied from some sample code).
  * Remove `// #region` folding support which doens't make sense in BQN.
  * Remove some characters from autoCloseBefore which looked like they didn't belong (based on the docs, "typically the set of characters which can not start an expression").
  * **Use Ctrl+Enter for both "Execute Line and Advance" and "Execute Selection"**, based on whether text is selected or not. There is no need for separate shortcuts (also, Ctrl+E which was used before conflicts with the default shortcut for jumping to the end of the line, which I use heavily since it's supported everywhere in macOS).
  * Remove the "cbqn" default, since "bqn.executablePath" already is already defaulted to "bqn" later in package.json.
  * Remove the "terminal.integrated.enablePersistentSessions" setting.
- v0.0.4
  - Multiline string support
  - Transient terminal
- Initial release
  - basic highlighting


gen_help.sh
#!/bin/bash

set -eufo pipefail

default_out=src/help.json

usage() {
    cat <<EOS
Usage: $(basename "$0") [-c] BQN_REPO_DIR [OUT_FILE]

Generates $default_out from Markdown files in the BQN repo.
Also checks that all URLs give 200 when -c is provided.
Used to produce help messages on hover in the VS Code extension.
EOS
}

check_urls=false
case ${1-} in
    -h|--help|help) usage; exit ;;
    -c) check_urls=true; shift ;;
esac

if [[ $# -eq 0 ]]; then
    usage >&2
    exit 1
fi

bqn_repo=$1
out_file=${2-$default_out}
# Useful when adding debug prints in the Lua file.
[[ "$out_file" == - ]] && out_file=/dev/stdout

if ! command -v pandoc &>/dev/null; then
    echo >&2 "error: pandoc must be installed"
    exit 1
fi

keymap_file="$(mktemp)"
url_file=$(mktemp)
trap 'rm -f "$keymap_file" "$url_file"' EXIT

node -e '
const obj = JSON.parse(
    require("fs").readFileSync("snippets/snippets.code-snippets")
);
for (const {body: [glyph], prefix: [p]} of Object.values(obj)) {
    if (p.startsWith("\\")) {
        console.log(glyph + p.slice(1));
    }
}
' > "$keymap_file"

find "$bqn_repo/help" -maxdepth 1 -name "*.md" -not -name "README.md" \
    | sort \
    | xargs pandoc -M keymap_file="$keymap_file" -M url_file="$url_file" \
        -f commonmark -t gen_help_writer.lua \
    | npx prettier  --parser json --no-config --tab-width 4 \
    > "$out_file"

if [[ $check_urls == true ]]; then
    echo >&2 "checking urls"
    n=0
    while read -r url; do
        {
            if ! curl --head --silent --fail "$url" 2>/dev/null; then
                echo >&2 "$url: not found"
                exit 1
            fi
            echo -n >&2 .
        } &
        ((n++))
        if [[ n -eq 10 ]]; then
            wait
        fi
    done < "$url_file"
    wait
    echo >&2
fi


src/extension.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode_1 = __importDefault(require("vscode"));
let terminal;
function activate(context) {
    const bqk = Array.from('`123456890-=~!@#$%^&*()_+qwertuiop[]QWERTIOP{}asdfghjkl;ASFGHKL:"zxcvbm,./ZXVBM<>? \'');
    const bqv = Array.from('˜˘¨⁼⌜´˝∞¯•÷×¬⎉⚇⍟◶⊘⎊⍎⍕⟨⟩√⋆⌽𝕨∊↑∧⊔⊏⊐π←→↙𝕎⍷𝕣⍋⊑⊒⍳⊣⊢⍉𝕤↕𝕗𝕘⊸∘○⟜⋄↖𝕊𝔽𝔾«⌾»·˙⥊𝕩↓∨⌊≡∾≍≠⋈𝕏⍒⌈≢≤≥⇐‿↩');
    let key_map = {};
    bqk.forEach((k, i) => key_map[k] = bqv[i]);
    // taken and modified from prollings/apl_backtick_symbols
    let pending = false;
    const command = vscode_1.default.commands.registerTextEditorCommand("language-bqn.backslash", (te, e) => {
        e.insert(te.selection.active, "\\");
        if (pending) {
            return 0;
        }
        pending = true;
        const apos = te.selection.active;
        const sub1 = vscode_1.default.workspace.onDidChangeTextDocument(_ => {
            sub1.dispose();
            const sub2 = vscode_1.default.workspace.onDidChangeTextDocument(ev => {
                sub2.dispose();
                const tpos = ev.contentChanges[0].range.start;
                const cond = (tpos.line === apos.line
                    && (tpos.character - apos.character) === 1
                    && ev.contentChanges[0].text.length === 1);
                if (cond) {
                    let range = new vscode_1.default.Range(apos, tpos.translate(0, 1));
                    let key = ev.contentChanges[0].text;
                    if (key in key_map) {
                        te.edit((e) => e.replace(range, key_map[key])).then();
                    }
                }
                pending = false;
            });
        });
    });
    const cmds = [
        ['language-bqn.createTerminal', createTerminal],
        ['language-bqn.loadScript', loadScript],
        ['language-bqn.executeSelection', executeSelection],
        ['language-bqn.executeLine', executeLine],
        ['language-bqn.executeLineAdvance', executeLineAdvance]
    ];
    for (const [n, f] of cmds) {
        vscode_1.default.commands.registerTextEditorCommand(n, f);
    }
    const tokenTypes = ['string'];
    const tokenModifiers = ['string'];
    const legend = new vscode_1.default.SemanticTokensLegend(tokenTypes, tokenModifiers);
    const provider = {
        provideDocumentSemanticTokens(document) {
            // analyze the document and return semantic tokens
            console.log(document);
            const tokensBuilder = new vscode_1.default.SemanticTokensBuilder(legend);
            // on line 1, characters 1-5 are a class declaration
            tokensBuilder.push(new vscode_1.default.Range(new vscode_1.default.Position(1, 1), new vscode_1.default.Position(1, 5)), 'class', ['declaration']);
            return tokensBuilder.build();
        }
    };
    const selector = { language: 'bqn', scheme: 'file' };
    const prov = vscode_1.default.languages.registerDocumentSemanticTokensProvider(selector, provider, legend);
    context.subscriptions.push(prov);
    context.subscriptions.push(command);
}
exports.activate = activate;
function deactivate(context) {
    if (terminal != null) {
        terminal.dispose();
    }
}
exports.deactivate = deactivate;
function createTerminal() {
    if (terminal == null || terminal.exitStatus != undefined) {
        const config = vscode_1.default.workspace.getConfiguration('bqn');
        terminal = vscode_1.default.window.createTerminal({
            name: "BQN", shellPath: config.executablePath
        });
        terminal.show();
    }
}
function loadScript(t, e) {
    createTerminal();
    terminal.sendText(`)ex ${t.document.fileName}`);
}
function executeSelection(t, e) { }
function executeLine(t, e) {
    createTerminal();
    let line = t.document.lineAt(t.selection.active.line).text;
    terminal.sendText(line, !line.endsWith('\n'));
}
function executeLineAdvance(t, e) {
    executeLine(t, e);
    vscode_1.default.commands.executeCommand('cursorMove', { to: "down", by: "wrappedLine" });
}


gen_help_writer.lua
-- Removes leading and trailing spaces from a string.
local function trim(s)
  return s:match "^%s*(.-)%s*$"
end

-- Returns `s` as a double-quoted JSON string.
local function json_str(s)
  return '"' .. s:gsub("[\\\"]", "\\%0"):gsub("\n", "\\n") .. '"'
end

-- Renders a list of Pandoc elements to CommonMark.
local function render(els)
  return trim(pandoc.write(pandoc.Pandoc(els), "commonmark", {
    wrap_text = "wrap-none"
  }))
end

-- Builds a table mapping BQN glyphs to the character used to type them after a
-- backslash, reading from a file where each line contains those two characters.
local function build_keymap(filename)
  local map = {}
  for line in io.lines(filename) do
    -- Note that we can't use indices 1 and 2 because the Unicode characters in
    -- the first column are more than one byte long.
    map[line:sub(1, -2)] = line:sub(-1)
  end
  return map
end

-- Converts a relative link to an absolute one, if it isn't already.
local function absolute_link(url)
  if url:match "^https://" then return url end
  local path, frag
  path, frag = url:match "^%.%./([^.]*)%.md(.*)$"
  if path then
    return "https://mlochbaum.github.io/BQN/" .. path .. ".html" .. frag
  end
  path, frag = url:match "^([^.]*)%.md(.*)$"
  if path then
    return "https://mlochbaum.github.io/BQN/help/" .. path .. ".html" .. frag
  end
  assert(false, "cannot convert url: " .. url)
end

-- Converts relative links to absolute in a Doc, and write links to url_file.
local function expand_relative_links(el, url_file)
  local written_urls = {}
  local f = assert(io.open(url_file, "w"))
  el = el:walk {
    Link = function(el)
      el.target = absolute_link(el.target)
      local url = el.target:gsub("#.*$", "")
      if not written_urls[url] then
        written_urls[url] = true
        f:write(el.target .. "\n")
      end
      return el
    end
  }
  f:close()
  return el
end

-- Parses the link from an element, or nil if there is none.
local function parse_link(el)
  local href
  el:walk {Link = function(el) href = el.target end}
  return href
end

-- Parses the "View this file with results and syntax highlighting here" link
-- that appears at the top of each file before the h1.
local function parse_help_page_link(el)
  local href = assert(parse_link(el))
  assert(pandoc.utils.stringify(el) == "View this file with results and syntax highlighting here.")
  return assert(href:match "^https://mlochbaum%.github%.io/BQN/help/.-%.html$", href)
end

-- Parses the "→full documentation" link that appears after most h2 elements,
-- or nil if there is none.
local function parse_full_documentation_link(el)
  local href = parse_link(el)
  if not href then return nil end
  local str = pandoc.utils.stringify(el)
  assert(str == "→full documentation" or str == "→specification", str)
  return href
end

-- Parses an h1 header, returning a list of {glyph, inlines} tables. The only
-- time it returns more than one is for Comma and Diamond as they share a page.
local function parse_h1(h1)
  local title = {}
  for i = 1, #h1.content do
    local el = h1.content[i]
    if el.t == "Str" and el.text == "(" then
      if i ~= #h1.content - 2 then
        local str = pandoc.utils.stringify(h1)
        assert(str == "Comma (,) and Diamond (⋄)", str)
        return {
          {glyph = ",", title = {pandoc.Str "Comma"}},
          {glyph = "⋄", title = {pandoc.Str "Diamond"}},
        }
      end
      local el1 = h1.content[i+1]
      local el2 = h1.content[i+2]
      assert(el1.t == "Code", el1.t)
      assert(el2.t == "Str", el2.t)
      assert(el2.text == ")", el2.text)
      local glyph = el1.text
      assert(#title > 0, #title)
      assert(title[#title].t == "Space", title[#title].t)
      table.remove(title)
      return {{glyph = glyph, title = title}}
    end
    table.insert(title, el)
  end
  assert(false, "no '(' found in h1")
end

-- Parses an h2 header, returning  to_colon and after_colon (lists of inlines).
local function parse_h2(h2)
  for i = 1, #h2.content do
    if h2.content[i].t == "Str" and h2.content[i].text == ":" then
      assert(i < #h2.content, i)
      assert(h2.content[i+1].t == "Space", h2.content[i+1].t)
      return {table.unpack(h2.content, 1, i+1)}, {table.unpack(h2.content, i+2)}
    end
  end
end

-- Returns true if `el` is a header with the given level.
local function is_header(el, level)
  return el.t == "Header" and el.level == level
end

function Writer(doc, options)
  local keymap = build_keymap(assert(doc.meta.keymap_file))
  doc = expand_relative_links(doc, assert(doc.meta.url_file))

  local first = true
  local headers, help_page_link, sections
  local function flush()
    if not headers then return end
    for _, header in ipairs(headers) do
      if not first then print "," end
      first = false
      print(json_str(header.glyph) .. ": [")
      local items = {
        pandoc.Link(header.title, help_page_link),
        pandoc.Space(),
        pandoc.Str '(',
        pandoc.Code(header.glyph),
      }
      local keymap_char = keymap[header.glyph]
      if keymap_char then
        table.insert(items, pandoc.Str ',')
        table.insert(items, pandoc.Space())
        table.insert(items, pandoc.Emph{pandoc. Str 'keymap:'})
        table.insert(items, pandoc.Space())
        table.insert(items, pandoc.Code("\\" .. keymap_char))
      end
      table.insert(items, pandoc.Str ')')
      print(json_str(render(pandoc.Para(items))))
      for i, section in ipairs(sections) do
        print("," .. json_str(render(section)))
      end
      print "]"
    end
    headers = nil
    help_page_link = nil
    sections = nil
  end

  print "{"
  local next_help_page_link, full_documentation_link
  local i = 1
  while i < #doc.blocks do
    local el = doc.blocks[i]
    if is_header(el, 1) then
      flush()
      headers = parse_h1(el)
      help_page_link = assert(next_help_page_link)
      next_help_page_link = nil
      assert(i < #doc.blocks, i)
      if doc.blocks[i+1].t == "Para" then
        -- The only time the "→full documentation" appears after the h1 instead
        -- of after the h2 is in beginexpression.md which documents `(`.
        assert(#headers == 1, #headers)
        assert(headers[1].glyph == "(", headers[1].glyph)
        full_documentation_link = assert(parse_full_documentation_link(doc.blocks[i+1]))
        i = i + 1
      end
      sections = {}
    elseif is_header(el, 2) then
      local inlines = el.content
      if not full_documentation_link then
        full_documentation_link = parse_full_documentation_link(doc.blocks[i+1])
      end
      if full_documentation_link then
        i = i + 1
        local after_colon
        inlines, after_colon = parse_h2(el)
        table.insert(inlines,
          pandoc.Link(pandoc.Strong(after_colon), full_documentation_link))
        full_documentation_link = nil
      else
        inlines = {pandoc.Strong(inlines)}
      end
      table.insert(sections, {pandoc.Para(inlines)})
    elseif is_header(el, 3) then
      -- Only nothing.md contains level 3 headers.
      assert(#headers == 1, #headers)
      assert(headers[1].glyph == "·", headers[1].glyph)
      assert(i < #doc.blocks, i)
      href = assert(parse_full_documentation_link(doc.blocks[i+1]))
      i = i + 1
      table.insert(sections, {pandoc.Para{pandoc.Link(el.content, href)}})
    elseif el.t == "Para" then
      if el.t == "Para" and i < #doc.blocks and is_header(doc.blocks[i+1], 1) then
        next_help_page_link = assert(parse_help_page_link(el))
      else
        assert(sections)
        assert(#sections > 0, #sections)
        table.insert(sections[#sections], el)
      end
    elseif el.t == "BulletList" then
      assert(sections)
      assert(#sections > 0, #sections)
      -- Example code blocks start with 8 spaces (4 spaces for the code block, 4
      -- for the BQN prompt). This extra indendation makes Pandoc think its part
      -- of the previous list item, so we strip them out here.
      el = el:walk {CodeBlock = function(el) return {} end}
      table.insert(sections[#sections], el)
    elseif el.t == "CodeBlock" then
      -- Don't include examples since they would make the hover docs too long
      -- and we'd need to run bqn to include their results.
    else
      assert(false, "unhandled case: " .. el.t)
    end
    i = i + 1
  end
  flush()
  print "}"

  -- Prevent pandoc from trying to print the returned doc.
  os.exit(0)
end


snippetGenerator.js
let syncls={ v:"Value", f:"Function", m:"Modifier", d:"Modifier2", n:"Number", g:"Gets", p:"Paren", b:"Bracket", k:"Brace", l:"Ligature", n:"Nothing", s:"Separator", c:"Comment", a:"String" };
let keydesc='f+Conjugate;Add_f-Negate;Subtract_f×Sign;Multiply_f÷Reciprocal;Divide_f⋆Exponential;Power_f√Square Root;Root_f⌊Floor;Minimum_f⌈Ceiling;Maximum_f∧Sort Up;And_f∨Sort Down;Or_f¬Not;Span_f|Absolute Value;Modulus_f≤Less Than or Equal to_f<Enclose;Less Than_f>Merge;Greater Than_f≥Greater Than or Equal to_f=Rank;Equals_f≠Length;Not Equals_f≡Depth;Match_f≢Shape;Not Match_f⊣Identity;Left_f⊢Identity;Right_f⥊Deshape;Reshape_f∾Join;Join to_f≍Solo;Couple_f↑Prefixes;Take_f↓Suffixes;Drop_f↕Range;Windows_f«Shift Before_f»Shift After_f⌽Reverse;Rotate_f⍉Transpose;Reorder axes_f/Indices;Replicate_f⍋Grade Up;Bins Up_f⍒Grade Down;Bins Down_f⊏First Cell;Select_f⊑First;Pick_f⊐Classify;Index of_f⊒Occurrence Count;Progressive Index of_f∊Mark First;Member of_f⍷Deduplicate;Find_f⊔Group Indices;Group_f!Assert;Assert with message_m˙Constant_m˜Self/Swap_d∘Atop_d○Over_d⊸Before/Bind_d⟜After/Bind_d⌾Under_d⊘Valences_d◶Choose_d⎉Rank_m˘Cells_d⚇Depth_m¨Each_m⌜Table_d⍟Repeat_m⁼Undo_m´Fold_m˝Insert_m`Scan_g←Define_g⇐Export_g↩Change_g→Return_s⋄Separator_s,Separator_p(Begin expression_p)End expression_k{Begin block_k}End block_b⟨Begin list_b⟩End list_l‿Strand_n·Nothing_v•System_v𝕨Left argument_f𝕎Left argument (as function)_v𝕩Right argument_f𝕏Right argument (as function)_v𝕗Modifier left operand (as subject)_f𝔽Modifier left operand_v𝕘2-modifier right operand (as subject)_f𝔾2-modifier right operand_v𝕤Current function (as subject)_f𝕊Current function_m𝕣Current modifier_n¯Minus_nπPi_n∞Infinity_a@Null character_c#Comment'.split(/[\n_]/);
let kk=Array.from('`123456890-=~!@#$%^&*()_+qwertuiop[]QWERTIOP{}asdfghjkl;ASFGHKL:"zxcvbm,./XVBM<>? \'');
let kv=Array.from('˜˘¨⁼⌜´˝∞¯•÷×¬⎉⚇⍟◶⊘⎊⍎⍕⟨⟩√⋆⌽𝕨∊↑∧⊔⊏⊐π←→↙𝕎⍷𝕣⍋⊑⊒⍳⊣⊢⍉𝕤↕𝕗𝕘⊸∘○⟜⋄↖𝕊𝔽𝔾«⌾»·˙⥊𝕩↓∨⌊≡∾≍≠𝕏⍒⌈≢≤≥⇐‿↩');
let keys={}, revkeys={}, primhelp={};
kk.map((k,i)=>{keys[k]=kv[i];revkeys[kv[i]]=k;});
let x = keydesc.map(d=>{
  let s = syncls[d[0]];
  let c = Array.from(d)[1];
  let t = d.slice(1+c.length).replace(";",", ");
  let k = revkeys[c];
  primhelp[c] = t;

  return [t,s,c,k];
});
//console.log(JSON.stringify(revkeys));

let shorts={}
for(let i of x){
a=i[1]+": "+i[0];
	shorts[a]={};
	shorts[a]["body"] = [i[2]];
	if(i[3]){
	shorts[a]["prefix"] = ["\\"+i[3]];
	} else {
	shorts[a]["prefix"] = [i[2]];
}
}
console.log(JSON.stringify(shorts));


src/extension.ts
import * as vscode from "vscode";
import * as path from "path";
import * as helpJson from "./help.json";

export function activate(context: vscode.ExtensionContext) {
  setupCmdBackslash(context);

  const commands = {
    "language-bqn.backslash": cmdBackslash,
    "language-bqn.createTerminal": cmdCreateTerminal,
    "language-bqn.loadScript": cmdLoadScript,
    "language-bqn.clearImports": cmdClearImports,
    "language-bqn.clearImportsAndLoadScript": cmdClearImportsAndLoadScript,
    "language-bqn.profileScript": cmdProfileScript,
    "language-bqn.executeSelection": cmdExecuteSelection,
    "language-bqn.executeLine": cmdExecuteLine,
    "language-bqn.executeLineAdvance": cmdExecuteLineAdvance,
  };
  for (const [name, callback] of Object.entries(commands)) {
    context.subscriptions.push(
      vscode.commands.registerTextEditorCommand(name, callback)
    );
  }

  if (getConfig().enableHoverDocumentation) {
    context.subscriptions.push(
      vscode.languages.registerHoverProvider("bqn", { provideHover })
    );
  }
}

export function deactivate() {
  if (terminal != undefined) {
    terminal.dispose();
  }
}

let pendingBackslashDecoration: vscode.TextEditorDecorationType;
const backslashKeymap: { [key: string]: string } = {};

function setupCmdBackslash(context: vscode.ExtensionContext) {
  const bqk = Array.from(
    "\\`123456890-=~!@#$%^&*()_+qwertuiop[]QWERTIOP{}asdfghjkl;ASFGHKL:\"zxcvbm,./ZXVBM<>? '"
  );
  const bqv = Array.from(
    "\\˜˘¨⁼⌜´˝∞¯•÷×¬⎉⚇⍟◶⊘⎊⍎⍕⟨⟩√⋆⌽𝕨∊↑∧⊔⊏⊐π←→↙𝕎⍷𝕣⍋⊑⊒⍳⊣⊢⍉𝕤↕𝕗𝕘⊸∘○⟜⋄↖𝕊𝔽𝔾«⌾»·˙⥊𝕩↓∨⌊≡∾≍≠⋈𝕏⍒⌈≢≤≥⇐‿↩"
  );
  for (const [i, key] of bqk.entries()) {
    backslashKeymap[key] = bqv[i];
  }
  pendingBackslashDecoration = vscode.window.createTextEditorDecorationType({
    backgroundColor:
      getConfig().pendingBackslashBackgroundColor ||
      new vscode.ThemeColor("editor.symbolHighlightBackground"),
  });
  context.subscriptions.push(pendingBackslashDecoration);
}

let backslashPending = false;

function cmdBackslash(editor: vscode.TextEditor, edit: vscode.TextEditorEdit) {
  const initialPositions = editor.selections.map((s) => s.active);
  for (const position of initialPositions) {
    edit.insert(position, "\\");
  }
  if (backslashPending) {
    return;
  }
  backslashPending = true;
  const subscription = vscode.workspace.onDidChangeTextDocument((event) => {
    for (const change of event.contentChanges) {
      if (!onChange(change)) {
        backslashPending = false;
        subscription.dispose();
        editor.setDecorations(pendingBackslashDecoration, []);
        break;
      }
    }
  });
  let unseenBackslashes = initialPositions.length;
  const editsToMake: [vscode.Range, string][] = [];
  const onChange = (change: vscode.TextDocumentContentChangeEvent): boolean => {
    if (unseenBackslashes > 0) {
      console.assert(change.text === "\\");
      unseenBackslashes--;
      if (unseenBackslashes === 0) {
        editor.setDecorations(
          pendingBackslashDecoration,
          initialPositions.map((p) => new vscode.Range(p, p.translate(0, 1)))
        );
      }
      return true;
    }
    const key = change.text;
    if (key.length !== 1) {
      return false;
    }
    const final = change.range.start;
    const initial = initialPositions.find((p) => {
      return p.line === final.line && p.character === final.character - 1;
    });
    if (initial == undefined) {
      return false;
    }
    const range = new vscode.Range(initial, final.translate(0, 1));
    const character = backslashKeymap[key];
    if (character == undefined) {
      return false;
    }
    editsToMake.push([range, character]);
    if (editsToMake.length !== initialPositions.length) {
      return true;
    }
    subscription.dispose();
    editor.edit((e) => {
      for (const [range, character] of editsToMake) {
        e.replace(range, character);
      }
    });
    return false;
  };
}

function provideHover(
  document: vscode.TextDocument,
  position: vscode.Position
) {
  const range = singleCharacterRange(document, position);
  const rawGlyph = document.getText(range);
  const glyph =
    { "𝕎": "𝕨", "𝕏": "𝕩", "𝕗": "𝔽", "𝕘": "𝔾", "𝕤": "𝕊" }[rawGlyph] ?? rawGlyph;
  const map = helpJson as { [glyph: string]: string[] };
  const help = map[glyph];
  if (help == undefined) {
    return;
  }
  return new vscode.Hover(help, range);
}

function singleCharacterRange(
  document: vscode.TextDocument,
  position: vscode.Position
): vscode.Range {
  const unitRange = new vscode.Range(position, position.translate(0, 1));
  const unit = document.getText(unitRange);
  if (unit >= "\uD800" && unit <= "\uDBFF") {
    // This is the first code unit of a UTF-16 surrogate pair. To get the full
    // code point, we have to include the next code unit as well.
    return new vscode.Range(position, position.translate(0, 2));
  }
  return unitRange;
}

let terminal: vscode.Terminal;
let terminalCwd: string;

async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function getConfig() {
  return vscode.workspace.getConfiguration("bqn");
}

async function createOrShowTerminal(
  editor: vscode.TextEditor,
  options: { preserveFocus: boolean }
) {
  if (terminal != undefined && terminal.exitStatus == undefined) {
    terminal.show(options.preserveFocus);
    return;
  }
  const config = getConfig();
  terminalCwd = path.dirname(editor.document.fileName);
  terminal = vscode.window.createTerminal({
    name: "BQN",
    shellPath: config.executablePath,
    location: vscode.TerminalLocation.Panel,
    cwd: terminalCwd,
  });
  terminal.show(options.preserveFocus);
  await sleep(config.sendToNewReplDelay);
}

function cmdCreateTerminal(editor: vscode.TextEditor) {
  createOrShowTerminal(editor, { preserveFocus: false });
}

async function runCommands(
  editor: vscode.TextEditor,
  commands: (script: string) => string[]
): Promise<void> {
  const tasks = [];
  tasks.push(createOrShowTerminal(editor, { preserveFocus: true }));
  const needsScriptParam = commands.length > 0;
  if (needsScriptParam && getConfig().saveBeforeLoadScript) {
    tasks.push(editor.document.save());
  }
  await Promise.all(tasks);
  // Use a relative path to save space, and make it more obvious when the user
  // is loading a script from a different directory than the one they started
  // the REPL in (and hence imports won't work properly).
  const script = path.relative(terminalCwd, editor.document.fileName);
  for (const cmd of commands(script)) {
    terminal.sendText(cmd, true);
  }
}

function cmdClearImports(editor: vscode.TextEditor) {
  runCommands(editor, () => [")clearImportCache"]);
}

function cmdLoadScript(editor: vscode.TextEditor) {
  runCommands(editor, (script) => [`)ex ${script}`]);
}

function cmdClearImportsAndLoadScript(editor: vscode.TextEditor) {
  runCommands(editor, (script) => [")clearImportCache", `)ex ${script}`]);
}

function cmdProfileScript(editor: vscode.TextEditor) {
  runCommands(editor, (script) => [`)profile ⟨⟩ •Import "${script}"`]);
}

async function execute(editor: vscode.TextEditor, text: string) {
  await createOrShowTerminal(editor, { preserveFocus: true });
  // We surround the text with control characters "\x1b[200~" and "\x1b[201~" to
  // indicate bracketed paste mode to replxx. This makes it treat the code as a
  // whole, rather than trying to evaluate after each newline, which would fail
  // e.g. for a line ending in "{" that starts a block.
  //
  // TODO: Once the extension API adds the third parameter `bracketedPasteMode`,
  // pass true for that instead of doing this manually. For more details:
  // https://github.com/microsoft/vscode/issues/153592#issuecomment-1182045382
  // https://github.com/microsoft/vscode/commit/65f97ed1d96989e64c57a689dae062579c75d5f3
  // https://github.com/microsoft/vscode/commit/89c00c59b6215802b8c02d9b0407f2a8340aed23
  if (getConfig().executableSupportsReplxx) {
    text = `\x1b[200~${text}\x1b[201~`;
  }
  terminal.sendText(text, true);
}

function cmdExecuteSelection(editor: vscode.TextEditor) {
  execute(editor, editor.document.getText(editor.selection));
}

function cmdExecuteLine(editor: vscode.TextEditor) {
  execute(editor, editor.document.lineAt(editor.selection.active.line).text);
}

function cmdExecuteLineAdvance(editor: vscode.TextEditor) {
  cmdExecuteLine(editor);
  const line = editor.selection.active.line;
  let n = 1;
  while (
    line + n < editor.document.lineCount - 1 &&
    /^\s*(#.*)?$/.test(editor.document.lineAt(line + n).text)
  ) {
    n++;
  }
  vscode.commands.executeCommand("cursorMove", {
    to: "down",
    by: "line",
    value: n,
  });
}
